### **Dataset** : Diabetes 130-US Hospitals for Years 1999–2008

### **Abou**t : It is a real-world healthcare dataset of 101,766 rows × 50 columns

In [1]:
!wget -O diabetes.zip "https://archive.ics.uci.edu/static/public/296/diabetes+130-us+hospitals+for+years+1999-2008.zip"
!unzip diabetes.zip

--2026-06-28 17:41:25--  https://archive.ics.uci.edu/static/public/296/diabetes+130-us+hospitals+for+years+1999-2008.zip
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified
Saving to: ‘diabetes.zip’

diabetes.zip            [ <=>                ]   3.02M  15.2MB/s    in 0.2s    

2026-06-28 17:41:26 (15.2 MB/s) - ‘diabetes.zip’ saved [3170254]

Archive:  diabetes.zip
  inflating: diabetic_data.csv       
  inflating: IDS_mapping.csv         


In [3]:
import os
print(os.listdir())

['.config', 'IDS_mapping.csv', 'diabetic_data.csv', 'diabetes.zip', 'sample_data']


### **Environment Setup and load the dataset**

In [4]:
import pandas as pd
import numpy as np

df = pd.read_csv("diabetic_data.csv")
df.head()

,encounter_id,patient_nbr,race,gender,age,weight,admission_type_id,discharge_disposition_id,admission_source_id,time_in_hospital,...,citoglipton,insulin,glyburide-metformin,glipizide-metformin,glimepiride-pioglitazone,metformin-rosiglitazone,metformin-pioglitazone,change,diabetesMed,readmitted
0,2278392,8222157,Caucasian,Female,[0-10),?,6,25,1,1,...,No,No,No,No,No,No,No,No,No,NO
1,149190,55629189,Caucasian,Female,[10-20),?,1,1,7,3,...,No,Up,No,No,No,No,No,Ch,Yes,>30
2,64410,86047875,AfricanAmerican,Female,[20-30),?,1,1,7,2,...,No,No,No,No,No,No,No,No,Yes,NO
3,500364,82442376,Caucasian,Male,[30-40),?,1,1,7,2,...,No,Up,No,No,No,No,No,Ch,Yes,NO
4,16680,42519267,Caucasian,Male,[40-50),?,1,1,7,1,...,No,Steady,No,No,No,No,No,Ch,Yes,NO


# **Basic Understanding regarding the dataset**

In [5]:
print(df.shape)

(101766, 50)


In [6]:
print(df.columns)

Index(['encounter_id', 'patient_nbr', 'race', 'gender', 'age', 'weight',
       'admission_type_id', 'discharge_disposition_id', 'admission_source_id',
       'time_in_hospital', 'payer_code', 'medical_specialty',
       'num_lab_procedures', 'num_procedures', 'num_medications',
       'number_outpatient', 'number_emergency', 'number_inpatient', 'diag_1',
       'diag_2', 'diag_3', 'number_diagnoses', 'max_glu_serum', 'A1Cresult',
       'metformin', 'repaglinide', 'nateglinide', 'chlorpropamide',
       'glimepiride', 'acetohexamide', 'glipizide', 'glyburide', 'tolbutamide',
       'pioglitazone', 'rosiglitazone', 'acarbose', 'miglitol', 'troglitazone',
       'tolazamide', 'examide', 'citoglipton', 'insulin',
       'glyburide-metformin', 'glipizide-metformin',
       'glimepiride-pioglitazone', 'metformin-rosiglitazone',
       'metformin-pioglitazone', 'change', 'diabetesMed', 'readmitted'],
      dtype='object')


In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101766 entries, 0 to 101765
Data columns (total 50 columns):
 #   Column                    Non-Null Count   Dtype 
---  ------                    --------------   ----- 
 0   encounter_id              101766 non-null  int64 
 1   patient_nbr               101766 non-null  int64 
 2   race                      101766 non-null  object
 3   gender                    101766 non-null  object
 4   age                       101766 non-null  object
 5   weight                    101766 non-null  object
 6   admission_type_id         101766 non-null  int64 
 7   discharge_disposition_id  101766 non-null  int64 
 8   admission_source_id       101766 non-null  int64 
 9   time_in_hospital          101766 non-null  int64 
 10  payer_code                101766 non-null  object
 11  medical_specialty         101766 non-null  object
 12  num_lab_procedures        101766 non-null  int64 
 13  num_procedures            101766 non-null  int64 
 14  num_

In [8]:
# Missing values
df.isnull().sum()

,0
encounter_id,0
patient_nbr,0
race,0
gender,0
age,0
weight,0
admission_type_id,0
discharge_disposition_id,0
admission_source_id,0
time_in_hospital,0


In [9]:
# Statistical summary
df.describe()

,encounter_id,patient_nbr,admission_type_id,discharge_disposition_id,admission_source_id,time_in_hospital,num_lab_procedures,num_procedures,num_medications,number_outpatient,number_emergency,number_inpatient,number_diagnoses
count,1.017660e+05,1.017660e+05,101766.000000,101766.000000,101766.000000,101766.000000,101766.000000,101766.000000,101766.000000,101766.000000,101766.000000,101766.000000,101766.000000
mean,1.652016e+08,5.433040e+07,2.024006,3.715642,5.754437,4.395987,43.095641,1.339730,16.021844,0.369357,0.197836,0.635566,7.422607
std,1.026403e+08,3.869636e+07,1.445403,5.280166,4.064081,2.985108,19.674362,1.705807,8.127566,1.267265,0.930472,1.262863,1.933600
min,1.252200e+04,1.350000e+02,1.000000,1.000000,1.000000,1.000000,1.000000,0.000000,1.000000,0.000000,0.000000,0.000000,1.000000
25%,8.496119e+07,2.341322e+07,1.000000,1.000000,1.000000,2.000000,31.000000,0.000000,10.000000,0.000000,0.000000,0.000000,6.000000
50%,1.523890e+08,4.550514e+07,1.000000,1.000000,7.000000,4.000000,44.000000,1.000000,15.000000,0.000000,0.000000,0.000000,8.000000
75%,2.302709e+08,8.754595e+07,3.000000,4.000000,7.000000,6.000000,57.000000,2.000000,20.000000,0.000000,0.000000,1.000000,9.000000
max,4.438672e+08,1.895026e+08,8.000000,28.000000,25.000000,14.000000,132.000000,6.000000,81.000000,42.000000,76.000000,21.000000,16.000000


In [10]:
# Target variable distribution
print(df["readmitted"].value_counts())

readmitted
NO     54864
>30    35545
<30    11357
Name: count, dtype: int64


### **Overview of the dataset**
RangeIndex : 101766 Rows, 0 to 101765

Total columns : 50

there are two data types:
a. int64 (13)
b. object (37)

Memory Usage : 38.8+ MB

## **Cleaning Notes**
1. **encounter_id** : Check missing values and convert data type to int64.
2. **patient_nbr**: Check missing values and convert data type to int64.
3. **race**: Replace "?" with NaN, fill missing values using Mode, then convert into numerical values.
4. **gender**: Remove "Unknown/Invalid" values and convert Male/Female into 1 and 0.
5. **age**: Convert age ranges ([0-10), [10-20), ...) into numeric age values.
6.** weight **: Most values are missing (?), so remove this column.
7. **admission_type_id **: Already numerical. Check for missing values.
8. **discharge_disposition_id** : Already numerical. Check for missing values.
9. **admission_source_id**: Already numerical. Check for missing values.
10. **time_in_hospital** : Check for invalid values and convert data type if required.
11. **payer_code** : Large number of missing values (?), remove this column.
12. **medical_specialty** : Replace "?" with NaN; if too many missing values, remove the column.
13. **num_lab_procedures** : Check for missing values and invalid values.
14. **num_procedures **: Check for missing values.
15. **num_medications** : Check for missing values.
16. **number_outpatient** : Check for missing values.
17. **number_emergency** : Check for missing values.
18.**number_inpatient**: Check for missing values.
19. **diag_1** : Replace "?" with "Unknown" and handle missing values.
20. **diag_2**: Replace "?" with "Unknown" and handle missing values.
21. **diag_3** : Replace "?" with "Unknown" and handle missing values.
22. **max_glu_serum** : Convert categorical values into numerical values.
23. **A1Cresult**: Convert categorical values into numerical values.
24. Medication Columns (metformin, repaglinide, nateglinide, ..., insulin) : Convert No, Steady, Up, and Down into numerical values.
25. **change** : Convert No and Ch into numerical values.
26. **bdiabetesMed** : Convert No and Yes into numerical values.
27. **readmitted**: Convert target values (NO, >30, <30) into numerical labels (0, 1, 2).
28. **Final Check** : Verify missing values, duplicate rows, and data types before saving the cleaned dataset.

# **DATA CLEANING**:

# **1. encounter_id**

### Checked datatype

In [11]:
df["encounter_id"].dtype

dtype('int64')

### Check missing values

In [12]:
df["encounter_id"].isnull().sum()

np.int64(0)

### Convert datatype

In [13]:
df["encounter_id"] = df["encounter_id"].astype("int64")

# **2. patient_nbr**

### Checked missing values

In [14]:
df["patient_nbr"].isnull().sum()

np.int64(0)

### Step 2 : Convert datatype

In [15]:
df["patient_nbr"] = df["patient_nbr"].astype("int64")

# **3. race**

### Check unique values

In [16]:
df["race"].unique()

array(['Caucasian', 'AfricanAmerican', '?', 'Other', 'Asian', 'Hispanic'],
      dtype=object)

### Check null values

In [17]:
df["race"].isnull().sum()

np.int64(0)

### Replace missing values with Mode

In [20]:
import numpy as np

df["race"] = df["race"].replace("?", np.nan)

mode_race = df["race"].mode()[0]

df["race"] = df["race"].fillna(mode_race)

### Convert categorical into numerical

In [21]:
race_map = {
    "Caucasian":1,
    "AfricanAmerican":2,
    "Asian":3,
    "Hispanic":4,
    "Other":5
}

df["race"] = df["race"].map(race_map)

# **4. gender**

### Check unique values

In [22]:
df["gender"].unique()

array(['Female', 'Male', 'Unknown/Invalid'], dtype=object)

### Remove invalid value

In [23]:
df = df[df["gender"] != "Unknown/Invalid"]

### Convert into numerical

In [24]:
gender_map = {
    "Male":1,
    "Female":0
}

df["gender"] = df["gender"].map(gender_map)

# **5. age**

### Check unique values

In [25]:
df["age"].unique()

array(['[0-10)', '[10-20)', '[20-30)', '[30-40)', '[40-50)', '[50-60)',
       '[60-70)', '[70-80)', '[80-90)', '[90-100)'], dtype=object)

### Convert age groups into numbers

In [26]:
age_map = {
    "[0-10)":5,
    "[10-20)":15,
    "[20-30)":25,
    "[30-40)":35,
    "[40-50)":45,
    "[50-60)":55,
    "[60-70)":65,
    "[70-80)":75,
    "[80-90)":85,
    "[90-100)":95
}

df["age"] = df["age"].map(age_map)

# **6. admission_type_id**

Already numeric

In [27]:
df["admission_type_id"].dtype

dtype('int64')

# **7. discharge_disposition_id**

Already numeric

In [28]:
df["discharge_disposition_id"].dtype

dtype('int64')

# **8. admission_source_id**

Already numeric

In [29]:
df["admission_source_id"].dtype

dtype('int64')

# **9. diagnosis columns (diag_1, diag_2, diag_3)**

### Replace missing values

In [30]:
df["diag_1"] = df["diag_1"].replace("?", np.nan)
df["diag_1"] = df["diag_1"].fillna("Unknown")

In [31]:
df["diag_2"] = df["diag_2"].replace("?", np.nan)
df["diag_2"] = df["diag_2"].fillna("Unknown")

In [32]:
df["diag_3"] = df["diag_3"].replace("?", np.nan)
df["diag_3"] = df["diag_3"].fillna("Unknown")

# **10. Binary columns**

In [33]:
binary_map = {
    "No":0,
    "Yes":1
}

df["change"] = df["change"].map(binary_map)

df["diabetesMed"] = df["diabetesMed"].map(binary_map)

# **11. Target Column**

In [34]:
df["readmitted"].unique()

array(['NO', '>30', '<30'], dtype=object)

### Convert into numbers

In [35]:
target_map = {
    "NO":0,
    ">30":1,
    "<30":2
}

df["readmitted"] = df["readmitted"].map(target_map)